keps/sig-api-machinery/storage-migration-auto-trigger.md

```diff
@@ -0,0 +1,229 @@
+---
+title: Automated Storage Version Migration with Storage Version Hash
+authors:
+  - "@xuchao"
+owning-sig: sig-api-machinery
+reviewers:
+  - "@deads2k"
+  - "@yliaog"
+approvers:
+  - "@deads2k"
+  - "@lavalamp"
+creation-date: 2019-01-23
+last-updated: 2019-01-23
+status: provisional
+---
+
+## Table of Contents
+
+* [Goal](#goal)
+* [API design](#api-design)
+* [Storage migration triggering controller](#storage-migration-triggering-controller)
+* [Implications to cluster operators](#implications-to-cluster-operators)
+* [Lifecycle of a StorageState object](#lifecycle-of-a-storagestate-object)
+* [Future work: HA clusters](#future-work-ha-clusters)
+* [Future work: persisted discovery document](#future-work-persisted-discovery-document)
+
+## Goal
+
+As the discovery document now exposes the [storage version hash][], storage version
+migration should start automatically when the hash changes.
+
+[storage version hash]:https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/35-storage-version-hash.md
+
+## API design
+
+We introduce the `StorageState` API, as a CRD defined in the `migration.k8s.io`
+group.
+
+```golang
+// StorageState is the state of the storage, for a specific resource.
+type StorageState struct {
+  metav1.TypeMeta
+  // The name is "<resource>.<group>". The API validation will enforce it by
+  // comapring the name with the spec.resource.
+  metav1.ObjectMeta
+  Spec StorageStateSpec
+  Status StorageStateStatus
+}
+
+type StorageStateSpec {
+  // The resource this StorageState is about.
+  Resource GroupResource
+}
+
+type StorageStateStatus {
+  // The hash values of possible stored versions, the versions to which the
+  // persisted instances of spec.resource are encoded.
+  // "Unknown" is a valid value in the list, and is the default value.
+  // It is not safe to upgrade or downgrade to an apiserver binary that does not
+  // support all versions listed in this field, or if "Unknown" is listed.
+  // Once the storage version migration for this resource has completed, the
+  // value of this field is refined to only contain the storageVersionHash.
+  // Once the apiserver has changed the storage version, the new storage version
+  // is appended to the list.
+  // +optional
+  StoredVersionsHashes []string
```

Comment:
Maybe the names for this and the next field can be made more clear, e.g. CurrentVersionHashes and DefaultVersionHash?

Reply:
I'm trying to keep them consistent with the CRD [StoredVersions](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go#L305) field and the apiserver's --storage-versions [flag](https://github.com/kubernetes/kubernetes/blob/0fdb65efa0d6c7d9ca81f6e218e97824909dc4c3/pkg/kubeapiserver/options/storage_versions.go#L98). WDYT?

Reply:
I think consistency with those would confusing, since the flag actually sets the default, not the list of supported versions.

The CRD StoredVersions field is accurately named, but it doesn't need to be paired with a "default" field.

Maybe "KnownVersionHashes" (or "PossibleVersionHashes") and "CurrentVersionHash" would be better than my first suggestion?

Reply:
I renamed them to "PersistedStorageVersionHashes" and "CurrentVersionHash" and updated the golang comment. WDYT?